test/e2e/base/manifests.yaml

```diff
@@ -521,3 +521,110 @@ spec:
     allowedRoutes:
       kinds:
       - kind: UDPRoute
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: gateway-preserve-case-backend
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: go-server
+  namespace: gateway-preserve-case-backend
+data:
+  go.mod: |
+         module srvr
+         go 1.21.3
+         require (
+         	github.com/andybalholm/brotli v1.0.5 // indirect
+         	github.com/klauspost/compress v1.17.0 // indirect
+         	github.com/valyala/bytebufferpool v1.0.0 // indirect
+         	github.com/valyala/fasthttp v1.51.0 // indirect
+         )
+  go.sum: |
+         github.com/andybalholm/brotli v1.0.5 h1:8uQZIdzKmjc/iuPu7O2ioW48L81FgatrcpfFmiq/cCs=
+         github.com/andybalholm/brotli v1.0.5/go.mod h1:fO7iG3H7G2nSZ7m0zPUDn85XEX2GTukHGRSepvi9Eig=
+         github.com/klauspost/compress v1.17.0 h1:Rnbp4K9EjcDuVuHtd0dgA4qNuv9yKDYKK1ulpJwgrqM=
+         github.com/klauspost/compress v1.17.0/go.mod h1:ntbaceVETuRiXiv4DpjP66DpAtAGkEQskQzEyD//IeE=
+         github.com/valyala/bytebufferpool v1.0.0 h1:GqA5TC/0021Y/b9FG4Oi9Mr3q7XYx6KllzawFIhcdPw=
+         github.com/valyala/bytebufferpool v1.0.0/go.mod h1:6bBcMArwyJ5K/AmCkWv1jt77kVWyCJ6HpOuEn7z0Csc=
+         github.com/valyala/fasthttp v1.51.0 h1:8b30A5JlZ6C7AS81RsWjYMQmrZG6feChmgAolCl1SqA=
+         github.com/valyala/fasthttp v1.51.0/go.mod h1:oI2XroL+lI7vdXyYoQk03bXBThfFl2cVdIA3Xl7cH8g=
+  main.go: |
+         package main
+         import (
+         	"encoding/json"
+         	"fmt"
+         	"log"
+         	"github.com/valyala/fasthttp"
+         )
+         func HandleFastHTTP(ctx *fasthttp.RequestCtx) {
+         	ctx.QueryArgs().VisitAll(func(key, value []byte) {
+         		if string(key) == "headers" {
+         			ctx.Response.Header.Add(string(value), "PrEsEnT")
+         		}
+         	})
+         	headers := map[string][]string{}
+         	ctx.Request.Header.VisitAll(func(key, value []byte) {
+         		headers[string(key)] = append(headers[string(key)], string(value))
+         	})
+         	if d, err := json.MarshalIndent(headers, "", "  "); err != nil {
+         		ctx.Error(fmt.Sprintf("%s", err), fasthttp.StatusBadRequest)
+         	} else {
+         		fmt.Fprintf(ctx, string(d)+"\n")
+         	}
+         }
+         func main() {
+         	s := fasthttp.Server{
+         		Handler:                       HandleFastHTTP,
+         		DisableHeaderNamesNormalizing: true,
+         	}
+            log.Printf("Starting on port 8000")
+         	log.Fatal(s.ListenAndServe(":8000"))
+         }
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: golang-app-deployment
+  namespace: gateway-preserve-case-backend
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: golang-app
+  template:
+    metadata:
+      labels:
+        app: golang-app
+    spec:
+      containers:
+      - name: golang-app-container
```

Comment:
this's a little trick, but it's clear to maintain the source code.

@arkodg WDYT?

Reply:
case preservation isn't possible with echo server today ?

Reply:
The echo server uses `net/http` which automatically and unconditionally normalizes headers with [CanonicalHeaderKey](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/http/request.go;l=165).

The minimal server I added uses [fasthttp](https://github.com/valyala/fasthttp) which supports not normalizing the headers.

The e2e test still isn't complete because while the server I added returns a header with non-standard casing, the `net/http` based client used in the test suite normalizes headers as well, so short of replacing the client side as well, it's not possible to add a check that returned headers' case is also preserved. (I did verify this manually using `curl` though.)

Reply:
@liorokman  This is brilliant :-) I think Ext Auth e2e test can aslo use it.

Reply:
cool thanks for explaining the current limitations @liorokman 
lets go ahead with this approach since @zirain and @zhaohuabing are +1s on this